File: dsa/SlidingWindow/maxSumSubArrOfK.py
## brute force

# nums = [2, 5, 1, 8, 2, 9, 1]
# n = len(nums)
# k = 3
# max = float('-inf')
# storage = []

# for i in range(n - k + 1):
#     sum = 0
#     for j in range(i, i + k):
#         sum += nums[j]
    
#     if max < sum:
#         max = sum
#     storage.append(sum)

# # print(max(storage))
# print(max)


# optimized solution : sliding window

nums = [2, 5, 1, 8, 2, 9, 1]
k = 3
n = len(nums)
max_val = float('-inf')
i = 0
j = 0
sum = 0
storage = []

while j < n:
    sum += nums [j]

    if j - i + 1 < k:
        j += 1

    elif j - i + 1 == k:

        if max_val < sum:
            max_val = sum
        
        storage.append(sum)
        sum -= nums[i]
        i += 1
        j += 1
    
print(storage, max_val)

File: dsa/SlidingWindow/maxSumSubArrOfK.js
// Maximum sum subarray of size k

// Brute force

// const nums = [2, 5, 1, 8, 2, 9, 1];
// const n = 7;
// const k = 3;
// const storage = [];
// let max = -Infinity;

// for (let i = 0; i < n - k + 1; i++) {
//   let sum = 0;
//   for (let j = i; j < i + k; j++) {
//     sum += nums[j];
//   }
//   storage.push(sum);
//   if (max < sum) {
//     max = sum;
//   }
// }
// console.log(max);

// optimized solution: sliding window

const nums = [2, 5, 1, 8, 2, 9, 1];
const n = 7;
const k = 3;
let i = 0;
let j = 0;
let sum = 0;
let max = -Infinity;
const storage = [];
while (j < n) {
  sum += nums[j];

  if (j - i + 1 < k) {
    j++;
  } else if (j - i + 1 === k) {
    if (max < sum) {
      max = sum;
    }
    storage.push(sum);
    sum -= nums[i];
    i++;
    j++;
  }
}
// console.log(max);
console.log(storage);
